designs/helper_to_upgrade_projects_by_rescaffolding.md

```diff
@@ -0,0 +1,212 @@
+`go/v3` to `go/v4`. The steps are very similar to the above ones. However,
+in this case you need to inform the plugin that you want to use to do the scaffold
+from scratch `kubebuilder alpha generate --plugins=go/v4`.
+
+## Open Questions
+
+N/A
+
+## Summary
+
+Therefore, a new command can be designed to load user configs from the [PROJECT config][project-config] file, and run the corresponding kubebuilder subcommands to generate the project based on the new kubebuilder version. Thus, it makes it easier for the users to migrate their operator projects to the new scaffolding.
+
+## Motivation
+
+A common scenario is to upgrade the project based on the newer Kubebuilder. The recommended (straightforward) steps are:
+
+- a) re-scaffold all files from scratch using the upper version/plugins
+- b) copy user-defined source code to the new layout
+
+The proposed command will automate the process at maximum, therefore helping operator authors with minimizing the manual effort.
+
+The main motivation of this proposal is to provide a helper for upgrades and
+make less painful this process. Examples:
+
+- See the discussion [How to regenerate scaffolding?](https://github.com/kubernetes-sigs/kubebuilder/discussions/2864)
+- From [slack channel By Paul Laffitte](https://kubernetes.slack.com/archives/CAR30FCJZ/p1675166014762669)
+
+### Goals
+
+- Help users upgrade their project with the latest changes
+- Help users to re-scaffold projects from scratch based on what was done previously with the tool
+- Make less painful the process to upgrade
+
+### Non-Goals
+
+- Change the default layout or how the KubeBuilder CLI works
+- Deal with customizations or deviations from the proposed layout
+- Be able to perform the project upgrade to the latest changes without human interactions
+- Deal and support external plugins
+- Provides support to [declarative](https://book.kubebuilder.io/plugins/declarative-v1.html) plugin
+  since it is desired and planned to decouple this solution and donate this plugin to its own authors [More info](https://github.com/kubernetes-sigs/kubebuilder/issues/3186)
+- Provide support to older version before having the Project config (Kubebuilder < 3x) and the go/v2 layout which exists to ensure  a backwards compatibility with legacy layout provided by Kubebuilder 2x
+
+## Proposal
+
+The proposed solution to achieve this goal is to create an alpha command as described
+in the example section above, see:
+
+```shell
+kubebuilder alpha generate \
+    --from=<path where the PROJECT file can be found> 
+    --to=<path where the project should be re-scaffold> 
+    --no-backup
+    --backup-path=<path-where the current version of the project should be copied as backup>
+    --plugins=<chain of plugins key that can be used to create the layout with init sub-command>
+```
+
+**Where**:
+
+- from: [Optional] If not informed then, by default it is the current directory (project directory). If the `PROJECT` file does not exist then it will fail.
```

Comment:
```suggestion
- input-dir: [Optional] If not informed then, by default it is the current directory (project directory). If the `PROJECT` file does not exist, then it will fail.
```